### ISSUE TITLE
Memory leak when calculating higher order gradients for graph using ReLU after Conv2d

### ISSUE THREAD

**Issue by vic4ever** [pytorch/pytorch#4245]
The following code causes memory leak each time backward is called. It only happens when Conv2D layer is followed by either ReLU or LeakyReLU layer. However, Conv2d then Tanh works fine. 

    import torch.autograd as autograd
    import gpustat
    from torch import optim
    import torch
    import torch.nn as nn

    class DCGANDiscriminator(nn.Module):
        def __init__(self, dim=64, bn=True):
            super(DCGANDiscriminator, self).__init__()
            self.dim = dim
            self.bn = bn

            self.conv1 = nn.Conv2d(3, 8*self.dim, 5,stride=2,padding=2,bias=True)
            #self.relu1 = nn.LeakyReLU(0.2)
            self.relu1 = nn.ReLU()

        def forward(self, input):
            output = input.contiguous()
            output = self.conv1(output)
            output = self.relu1(output)
            output = output.view(-1)
            return output

    def show_memusage(device=1):
        gpu_stats = gpustat.GPUStatCollection.new_query()
        item = gpu_stats.jsonify()["gpus"][device]
        print('Used/total: ' + "{}/{}".format(item["memory.used"], item["memory.total"]))

    aD = DCGANDiscriminator()
    aD = aD.cuda()
    for i in range(1000):
        print("Critic iter: " + str(i))
        aD.zero_grad()
        real_data = autograd.Variable(torch.randn(1,3,64,64),requires_grad=True).cuda()

        show_memusage(0)
        out = aD(real_data)
        gradients = autograd.grad(outputs=out, inputs=real_data,
                                  grad_outputs=torch.ones(out.size()).cuda(),
                                  create_graph=True, retain_graph=True, only_inputs=True)[0]
        gradients = gradients.view(gradients.size(0), -1)
        gradient_penalty = gradients.mean()
        gradient_penalty.backward()
        show_memusage(0)

Here is some output:

    Critic iter: 0
    Used/total: 1725/11172
    Used/total: 1835/11172
    Critic iter: 1
    Used/total: 1835/11172
    Used/total: 1837/11172
    Critic iter: 2
    Used/total: 1837/11172
    Used/total: 1839/11172
    Critic iter: 3
    Used/total: 1839/11172
    Used/total: 1841/11172
    Critic iter: 4
    Used/total: 1841/11172
    Used/total: 1843/11172
    Critic iter: 5
    Used/total: 1843/11172
    Used/total: 1845/11172
    Critic iter: 6
    Used/total: 1845/11172
    Used/total: 1847/11172

I'm using pytorch version 0.4.0a0+e2c75d3 on Ubuntu 16.04, python 3.6, CUDA 8.

**Comment by colesbury** [comment]
This should be fixed in master (as of d605058, which was merged yesterday).  Can you let us know if you're still seeing this behavior on the latest version?

**Comment by colesbury** [comment]
I'm not seeing a leak on master: https://gist.github.com/colesbury/708a7c714c64fab4611372a7b624d656

**Comment by vic4ever** [comment]
You are right. The commit fixed the bug. Thanks.
